### PR TITLE
Added not-ready exception

### PR DESCRIPTION
**Pull request by EranGon**

### CODE REVIEW

File: src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+// </copyright>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.Azure.Monitoring.SmartDetectors
+{
+    using System;
+    using System.Runtime.Serialization;
+
+    /// <summary>
+    /// An exception used by the Smart Detector to indicate that the data it need for analysis
```
Comment (YairoR) [inline src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs]: data it needs

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs [hunk 2]
```diff
@@ -153,7 +153,13 @@ private static string GetResolutionStateKey(string alertCorrelationHash)
...
             {
                 alerts = await smartDetector.AnalyzeResourcesAsync(analysisRequest, detectorTracer, cancellationToken);
-                this.tracer.TraceInformation($"Completed running Smart Detector ID {smartDetectorManifest.Id}, Name {smartDetectorManifest.Name}, returning {alerts.Count} alerts");
+                this.tracer.TraceInformation(
+                    $"Completed running Smart Detector ID {smartDetectorManifest.Id}, Name {smartDetectorManifest.Name}, returning {alerts.Count} alerts");
+            }
+            catch (DetectorDataNotReadyException ddnre)
+            {
+                this.tracer.TraceWarning($"Smart Detector data is not ready yet, aborting analysis: {ddnre.Message}");
+                return new List<ContractsAlert>();
```
Comment (YairoR) [inline src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: Is that temporary? otherwise how will you know to return it to the queue

Comment (EranGon) [reply src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: this is not supported in the Monitoring Appliance - it's an optional feature than may or may not be implemented by the Runtime Environment

File: src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    /// is not ready yet. The Runtime Environment may use <see cref="ShouldRetryAfter"/> to
+    /// reschedule the analysis, but in any case throwing this exception will not register as
+    /// a failure to run the Smart Detector.
+    /// </summary>
+    [Serializable]
+    public class DetectorDataNotReadyException : Exception
```
Comment (ilaizi) [inline src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs]: >DetectorDataNotReadyException [](start = 17, length = 29)

Why did you decide not to call it DetectorSnoozeException? Is it too informal?

File: src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs (hunk 1 above)
Comment (ilaizi) [inline src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs]: >need [](start = 77, length = 4)

needs

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj [hunk 4]
```diff
@@ -22,7 +22,7 @@
   <Import Project="$(NuProjPath)\NuProj.props" Condition="Exists('$(NuProjPath)\NuProj.props')" />
   <PropertyGroup Label="Configuration">
     <Id>Microsoft.Azure.Monitoring.SmartDetectors</Id>
-    <Version>0.14.0.0</Version>
+    <Version>0.14.1.0</Version>
```
Comment (ilaizi) [inline src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: If I remember correctly, the was an additional nuget.
Should we update its version also?

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs (hunk 2 above)
Comment (ilaizi) [reply src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: Can we add some kind of a flag to analysis request to let the user know if the runtime environment supports this?

---
In reply to: [243836908](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/38#discussion_r243836908) [](ancestors = 243836908)

File: src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs (hunk 3 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsSDK/DetectorDataNotReadyException.cs]: yes

---
In reply to: [243895026](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/38#discussion_r243895026) [](ancestors = 243895026)

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj (hunk 4 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: no - minor versions can live seperately

---
In reply to: [243895875](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/38#discussion_r243895875) [](ancestors = 243895875)

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs (hunk 2 above)
Comment (EranGon) [reply src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: i would rather not - the flag's value should not change the detector's decision in any case

---
In reply to: [243896042](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/38#discussion_r243896042) [](ancestors = 243896042,243836908)

**Pull request merged**